contracts/mixins/ProtocolTokenHandler.sol

```diff
@@ -51,51 +65,75 @@ contract ProtocolTokenHandler is ProtocolTokenUser {
 
     /**
      * @dev Before calling this function to deposit protocol tokens, 
-     * users need approve this contract to be able to spend or transfer their protocol tokens.    
+     * users need approve this contract to be able to spend or transfer their protocol tokens.
+     * @param _amount amount of protocol tokens deposited to this contract
      */
     function deposit(uint256 _amount) public {
         require(_amount > 0, "amount should larger than zero");
 
         IERC20(protocolTokenAddress).safeTransferFrom(msg.sender, address(this), _amount);
-        protocolTokenHeld = protocolTokenHeld.add(_amount);
 
         emit Deposit(msg.sender, _amount);
     }
 
     /**
-     * @param _data the message/data from users to be signed by authorized signers, should include in order: 
-     * address of recipient, addresss of destination contract (should be this contract address), amount of tokens to be withdrawed, nounce
      * @param _signs array includes signs from authorized signers, amount/length should equal to requiredCount
+     * @param _recipient address of recipient
+     * @param _contractAddress addresss of destination contract (should be this contract address)
+     * @param _amount amount of tokens to be withdrawed
+     * @param _nonce nonce to avoid replay attacks
+     * @param _signs array includes signs from authorized signers
      */
-    function withdraw(bytes memory _data, bytes[] memory _signs) public {
-        _signs = new bytes[](requiredCount);
-        bytes32 _hash = keccak256(_data).toEthSignedMessageHash();
+    function withdraw(address _recipient, address _contractAddress, uint256 _amount, uint256 _nonce, bytes[] memory _signs) public {
+        require(_contractAddress == address(this), "contract address disagreement");
+        require(_amount > 0, "amount should larger than zero");
 
-        for (uint256 i = 0; i < requiredCount; i++) {
-            address signer = _hash.recover(_signs[i]);
-            require(isSigner[signer], "signer not authorized");
-            require(!verifiedSigners[signer], "signer verified");
-            verifiedSigners[signer] = true;
-        }
+        uint256 oldNonce = userNonce[msg.sender];
+        require(_nonce > oldNonce, "nonce smaller than last know nonce");
```

Comment:
wukong, i'm sorry, because i requested this change, but i think your last solution was actually better fit for the purpose of this contract. could you rename nonce to taskID? and check that each taskID is only processed once per user? taskIDs can then be passed in any order. E.g. we can first process ID number 7, then ID number 3. but it has to be per user.